### ISSUE TITLE
Support none `nodata`

### ISSUE THREAD

**Issue by emmaai** [opendatacube/datacube-core#614]
### Expected behaviour
datacube should support geotiff or netcdf without `nodata`, aka, all the data are valid. 

### Actual behaviour
`nodata` value is compulsory in product definition. While the geotiff has data type `unit8` and all the data valid, the best can be done in product definition is to set `dtype=int16` and `nodata=-1`. In this case, the value `255` will be cast as `-1` even though `255` is essentially valid data.

**Comment by Kirill888** [comment]
When reading data, datacube assumes that some data is not valid:

https://github.com/opendatacube/datacube-core/blob/6c9b4d7ed19a8a240b54b781a5917dc87c3bb053/datacube/storage/storage.py#L162

Up there `src.nodata` is 
1. Read from file and converted to `dtype` of source pixel data
2. Or read from product definition and converted to `dtype` of source pixel data

https://github.com/opendatacube/datacube-core/blob/6c9b4d7ed19a8a240b54b781a5917dc87c3bb053/datacube/storage/storage.py#L354-L355
where `self.nodata` is "fallback value" read from product defintion:
https://github.com/opendatacube/datacube-core/blob/6c9b4d7ed19a8a240b54b781a5917dc87c3bb053/datacube/storage/storage.py#L404


Also when doing reproject `src_nodata` is overwritten with product version if not present in the file:

https://github.com/opendatacube/datacube-core/blob/6c9b4d7ed19a8a240b54b781a5917dc87c3bb053/datacube/storage/storage.py#L257

Basically Datacube behaviour when working with files that do no define `nodata` value is to assume that *"file is wrong, I know better"*, but we need to support *"file has valid pixels only"* case as well or instead.

**Comment by Kirill888** [comment]
Blindly converting product definition version of `nodata` value to the `dtype` of source image is a mistake. For example `-999: int16` will become `25: uint8` without any warning, and is certainly not what I would expect to see.

**Comment by Kirill888** [comment]
I have confirmed that both `reproject` and `read` from a `uint8` geotiff file that has no `nodata` value defined will replace `255` pixels with `-1`, when product defines `dtype=int16, nodata=-1`
